### PR TITLE
Add AutoScrollingBehavior with option to scroll to item on insert

### PR DESCRIPTION
**Pull request by kylebshr**
This PR adds an enum for configuring the list view with `AutoScrollingBehavior`. It has two cases: `none` and `scrollToItemOnInsert(_:, position:)`.

When the list is updated, if `scrollToItemOnInsert` is set, the list checks if the item in the enum case was added in this update. If it was, it scrolls to the item after reloading the content. You can see this in the attached GIF:

Kapture 2020-04-01 at 11 35 04
Fixes https://github.com/kyleve/Listable/issues/96

### CODE REVIEW

File: Listable/Sources/ListView/ListView.swift [hunk 1]
```diff
@@ -229,7 +229,47 @@ public final class ListView : UIView
...
+            }
+
+            let contentOffsetY = contentHeight - contentFrameHeight -
+                self.collectionView.lst_adjustedContentInset.top
+            let contentOffset = CGPoint(x: self.collectionView.contentOffset.x, y: contentOffsetY)
+            self.collectionView.setContentOffset(contentOffset, animated: animated)
+        }
+
+        if lastLoadedIndexPath < toIndexPath {
+            self.updatePresentationState(for: .programaticScrollDownTo(toIndexPath)) { _ in
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift]: Since this is duplicated across the scroll item and this method, can we centralize the "I need to update the slice" bit?

File: Listable/Sources/Behavior.swift [hunk 2]
```diff
@@ -11,9 +11,14 @@ import Foundation
 public struct Behavior : Equatable
 {
     public var dismissesKeyboardOnScroll : Bool
+    public var pinItemsToBottom : Bool
```
Comment (kyleve) [inline Listable/Sources/Behavior.swift]: staysPinnedToLastItem? staysScrolledToLastItem?

Comment (kylebshr) [reply Listable/Sources/Behavior.swift]: The more I think about this the more I think we should only scroll to the bottom if an item is _inserted_ at the bottom, right? If you delete an item at the bottom, the standard `performUpdates` will handle this, and if you insert/delete/move anything else, you probably don't want this to happen. If you agree, then maybe the name should reflect that - `scrollsToBottomOnInsertion` or something

File: Listable/Sources/ListView/ListView.swift [hunk 3]
```diff
@@ -577,6 +625,17 @@ public final class ListView : UIView
             self.updateVisibleItemsAndSections()
             callerCompletion(finished)
         }
+
+        // Perform any other behavior-related changes.
+
+        switch reason {
+        case .contentChanged:
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift]: Probably, I'll try that!

File: Listable/Sources/Behavior.swift (hunk 2 above)
Comment (kylebshr) [reply Listable/Sources/Behavior.swift]: `pinBottomInserts` `pinInsertionsToBottom`

Comment (kyleve) [reply Listable/Sources/Behavior.swift]: `pinsToBottomForNewItems`?

**Comment by kyleve** [comment]
@kylebshr is this worth reviewing again now?

File: Listable/Sources/Behavior.swift (hunk 2 above)
Comment (kyleve) [reply Listable/Sources/Behavior.swift:14]: Bump

File: Listable/Sources/ListView/ListView.swift [hunk 4]
```diff
@@ -217,23 +213,39 @@ public final class ListView : UIView
...
-            self.updatePresentationState(for: .programaticScrollDownTo(toIndexPath)) { _ in
-                scroll()
+    }
+
+    @discardableResult
+    public func scrollToBottom(animated : Bool = true) -> Bool {
+
+        // Make sure we have a valid last index path.
+
+        guard let toIndexPath = self.storage.allContent.lastIndexPath() else {
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:230]: We have this, and then we also have `lastIndexPath` on `PresentationState` – can the `PresentationState` one defer through to this?

File: Listable/Sources/ListView/ListView.swift [hunk 5]
```diff
@@ -521,6 +533,10 @@ public final class ListView : UIView
         case .contentChanged:
             self.updateCollectionViewConfiguration()
             self.updatePresentationStateWith(firstVisibleIndexPath: indexPath, for: reason, completion: completion)
+
+            if self.behavior.pinItemsToBottom {
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:537]: I think we should only do this if we inserted an item, right? I think we need to think through this behaviour a bit more. Eg, we may also want to remove the `Behaviour` level change, and instead tag items as `scrollToWhenInserted` or similar.

File: Listable/Sources/ListView/ListView.swift (hunk 4 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:230]: Probably! Will look into this

File: Listable/Sources/Behavior.swift (hunk 2 above)
Comment (kylebshr) [reply Listable/Sources/Behavior.swift:14]: Sorry for the long lapse on this, will be wrapping up some changes and pushing today! `pinsToBottomForNewItems` seems good to me

File: Listable/Sources/ListView/ListView.swift (hunk 5 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:537]: > I think we should only do this if we inserted an item, right?

Yep, agreed, though the two qualms I have with making it an item level change are:

- More work for the consumer when they just want pinToBottom, since they have to figure out which items to add the flag to (not actually that hard though)
- Scrolling to an item doesn't have the same behavior as pin to bottom, as it doesn't account for safe area and content insets

File: Listable/Sources/ListView/ListView.swift (hunk 4 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:244]: I think the reason I needed this is because `PresentationState` may not have all the items in it, so the `lastIndexPath` isn't the true last one of the list - is that assumption correct?

Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:244]: Ahh true

File: Listable/Sources/ListView/ListView.swift [hunk 6]
```diff
@@ -200,38 +203,58 @@ public final class ListView : UIView
         guard let toIndexPath = self.storage.allContent.indexPath(for: item) else {
             return false
         }
+
+        // Check if the item is visible using its frame, since `visibleIndexPaths` includes items outside of the actual content frame.
```
Comment (kylebshr) [inline Listable/Sources/ListView/ListView.swift:207]: This fixes a bug where scrolling to an item wouldn't work for `doNothing` if it was 1-2 items offscreen.

File: Listable/Sources/ListView/ListView.swift [hunk 7]
```diff
@@ -215,23 +227,39 @@ public final class ListView : UIView
...
             )
         }
-        
-        if lastLoadedIndexPath < toIndexPath {
-            self.updatePresentationState(for: .programaticScrollDownTo(toIndexPath)) { _ in
-                scroll()
+    }
+
+    @discardableResult
+    public func scrollToBottom(animated : Bool = true) -> Bool {
```
Comment (kylebshr) [inline Listable/Sources/ListView/ListView.swift:235]: Thinking we should keep this since it seems generally useful, should I move to another PR?

File: Listable/Sources/ListView/ListView.swift [hunk 8]
```diff
@@ -578,12 +614,39 @@ public final class ListView : UIView
             self.updateVisibleItemsAndSections()
             callerCompletion(finished)
         }
-        
+
+        if case let AutoScrollingBehavior.scrollToItemOnInsert(autoScrollItem, autoScrollPosition) = self.autoScrollingBehavior,
```
Comment (kylebshr) [inline Listable/Sources/ListView/ListView.swift:618]: I couldn't figure out a nice way to keep this in the switch above, since I need the switch to calculate the diff, and need the diff to do this. If anyone sees a more elegant solution let me know

File: Listable/Sources/ListDescription.swift [hunk 9]
```diff
@@ -14,6 +14,7 @@ public struct ListDescription
     
     public var appearance : Appearance
     public var behavior : Behavior
+    public var autoScrollingBehavior : AutoScrollingBehavior
```
Comment (kyleve) [inline Listable/Sources/ListDescription.swift:17]: Maybe `autoScrollAction`?

File: Listable/Sources/ListView/ListView.swift (hunk 6 above)
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:207]: Ooh, good catch

File: Listable/Sources/ListView/ListView.swift [hunk 10]
```diff
@@ -154,6 +155,8 @@ public final class ListView : UIView
     {
         // Nothing right now.
     }
+
+    public var autoScrollingBehavior : AutoScrollingBehavior
```
[the remark this answers is not in the document]
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:159]: (It's of course not bad to have it here, just thinking out loud)

File: Listable/Sources/ListView/ListView.swift [hunk 11]
```diff
@@ -553,11 +578,22 @@ public final class ListView : UIView
...
+        let visibleSlice: Content.Slice
+
+        if self.bounds.isEmpty {
+            visibleSlice = Content.Slice()
+        } else {
+            switch self.autoScrollingBehavior {
+            case .none:
+                visibleSlice = self.storage.allContent.sliceTo(indexPath: indexPath, plus: Content.Slice.defaultSize)
+            case .scrollToItemOnInsert:
+                visibleSlice = Content.Slice(containsAllItems: true, content: self.storage.allContent)
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:591]: This should probably only slice to the index of the item if it's contained, right?

File: Listable/Sources/ListView/ListView.swift [hunk 12]
```diff
@@ -578,12 +614,39 @@ public final class ListView : UIView
             self.updateVisibleItemsAndSections()
             callerCompletion(finished)
         }
-        
+
+        if case let AutoScrollingBehavior.scrollToItemOnInsert(autoScrollItem, autoScrollPosition) = self.autoScrollingBehavior,
+            !diff.old.contains(item: autoScrollItem), diff.new.contains(item: autoScrollItem)
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:619]: Wondering if it's easier / faster to instead check the inserted items in the diff – since this could be comparing a _lot_ of items.

File: Listable/Sources/ListView/ListView.swift [hunk 13]
```diff
@@ -578,12 +614,39 @@ public final class ListView : UIView
...
+        {
+            self.scrollTo(item: autoScrollItem, position: autoScrollPosition, animated: true)
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:621]: Should animated always be true here?

File: Listable/Sources/ListView/ListView.swift [hunk 14]
```diff
@@ -768,3 +847,10 @@ fileprivate extension UIScrollView
         }
     }
 }
+
+private extension Array where Element == Section {
+    func contains(item: AnyItem) -> Bool {
+        return self.flatMap { $0.items }
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:853]: Wondering about the performance of this for large sections / lists. What about a nested `contains(where:)` instead?

File: Listable/Sources/ListView/ListView.swift [hunk 15]
```diff
@@ -200,38 +203,58 @@ public final class ListView : UIView
...
+            guard contentHeight > contentFrameHeight else {
+                return
             }
-        } else {
-            scroll()
+
+            let contentOffsetY = contentHeight - contentFrameHeight -
+                self.collectionView.lst_adjustedContentInset.top
+            let contentOffset = CGPoint(x: self.collectionView.contentOffset.x, y: contentOffsetY)
+            self.collectionView.setContentOffset(contentOffset, animated: animated)
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:256]: `updatePresentationStateAndScroll` is probably a confusing method name.. it takes a function to do the scrolling, which is what is being implemented here. Will clean up.

File: Listable/Sources/ListView/ListView.swift (hunk 11 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:591]: ah yeah, meant to change

File: Listable/Sources/ListView/ListView.swift (hunk 12 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:619]: _Is_ there an inserted items I'm missing? I couldn't find one 🤔

File: Listable/Sources/ListView/ListView.swift (hunk 13 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:621]: Maaybe not I was thinking it would be but no reason not to make it configurable

File: Listable/Sources/ListView/ListView.swift (hunk 14 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:853]: Probably better!

File: Listable/Sources/ListView/ListView.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:159]: Alternative would be to pass through `setProperties` → `setContent` → `set` → `setContentFromSource` → `updatePresentationState` - property seems simpler to me

[the remark this answers is not in the document]
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift]: Sounds good!

File: Listable/Sources/ListView/ListView.swift (hunk 12 above)
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:620]: Kind of, but it's a level below where you'd expect because there is first a diff of sections which then contain changes to individual items. I was thinking something like this: https://github.com/kyleve/Listable/compare/kve/insertedIDs?expand=1

File: Listable/Sources/ListView/ListView.swift (hunk 13 above)
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:622]: Thinking eg, you set up a list with no animation because you're about to push it on screen, we wouldn't want this to have animation

File: Listable/Sources/ListView/ListView.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:256]: Ahhh got it, right

File: Listable/Sources/ListView/ListView.swift (hunk 14 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:854]: Will actually remove this in favor of checking in the diff like you suggested above

File: Listable/Sources/AutoScrollAction.swift [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+import Foundation
+
+
+public enum AutoScrollAction
+{
+    /// The list never automatically scrolls.
+    case none
+
+    /// Scrolls to the specified item when the list is updated.
```
Comment (kyleve) [inline Listable/Sources/AutoScrollAction.swift:16]: Nit: Comment should mention that it only happens on insert

File: Listable/Sources/ListView/ListView.swift [hunk 17]
```diff
@@ -553,11 +578,23 @@ public final class ListView : UIView
...
+
+        let visibleSlice: Content.Slice
+
+        if self.bounds.isEmpty {
+            visibleSlice = Content.Slice()
+        } else {
+            switch self.autoScrollAction {
+            case .none:
+                visibleSlice = self.storage.allContent.sliceTo(indexPath: indexPath, plus: Content.Slice.defaultSize)
+            case .scrollToItemOnInsert(let autoScrollItem, _):
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:590]: I don't think there's a great solution here, but it would be nice if we only use the `autoScrollItem`'s index path when performing the scroll. Wonder if `Content.Slice.UpdateReason` could be leveraged for this?

File: Listable/Sources/ListView/ListView.swift [hunk 18]
```diff
@@ -553,11 +578,23 @@ public final class ListView : UIView
...
+                let indexPath = self.storage.allContent.indexPath(for: autoScrollItem.identifier) ?? indexPath
+                visibleSlice = self.storage.allContent.sliceTo(indexPath: indexPath, plus: Content.Slice.defaultSize)
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:592]: This needs to be either this `indexPath`, OR the `firstVisibleIndexPath`, whichever is larger.

File: Listable/Sources/ListView/ListView.swift (hunk 12 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift]: Updated! Blatantly stole your code, but added some tests - I think it works as we need it to.

File: Listable/Sources/ListView/ListView.swift [hunk 19]
```diff
@@ -200,38 +203,58 @@ public final class ListView : UIView
...
+
+        // Make sure we have a valid last index path.
+
+        guard let toIndexPath = self.storage.allContent.lastIndexPath() else {
+            return false
+        }
+
+        // Perform scrolling.
+
+        return self.updatePresentationStateForScroll(toIndexPath: toIndexPath)  {
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:245]: Was thinking, maybe `extendPresentationState(to indexPath : IndexPath) { ... }`?

Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift:245]: Or `preparePresentationStateForScroll(to:)`?

File: Listable/Sources/ListView/ListView.swift (hunk 17 above)
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift]: Hmm, we could set it back to `none` if it was `scrollToItemOnInsert`, which means it would just use the `firstVisibleIndexPath` next time unless `autoScrollAction` was set again. Does that seem like a good fix?

File: Listable/Tests/Internal/Diff/SectionedDiffTests.swift [hunk 20]
```diff
@@ -13,3 +13,116 @@ class SectionedDiffTests: XCTestCase
...
+
+class SectionedDiff_SectionChangesTests: XCTestCase
+{
+
+    struct Section {
+        let identifier: Int
+        let items: [Int]
+    }
+
+    func test_added_items_in_added_section() {
```
Comment (kyleve) [inline Listable/Tests/Internal/Diff/SectionedDiffTests.swift:25]: Test structure nit: use `func testcase("description") { ... }` within a single test method to group related tests?

File: Listable/Sources/ListView/ListView.swift (hunk 17 above)
Comment (kyleve) [reply Listable/Sources/ListView/ListView.swift]: I think so – this is one reason I was thinking it doesnt make sense to store `autoScrollAction` as a property, since it really is more of an action to perform during the current update cycle vs. to persist on the view itself. Hmmm

Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift]: Yeah good point.. I'm gonna play with piping it through the functions

File: Listable/Sources/ListView/ListView.swift [hunk 21]
```diff
@@ -578,12 +621,39 @@ public final class ListView : UIView
...
+            self.scrollTo(item: autoScrollItem, position: autoScrollPosition, animated: animated)
+        }
+
         // Update info for new contents.
         
         self.updateCollectionViewSelections(animated: reason.animated)
     }
-        
+
+    private func preparePresentationStateForScroll(toIndexPath: IndexPath, scroll: @escaping () -> Void) -> Bool {
```
Comment (kyleve) [inline Listable/Sources/ListView/ListView.swift:636]: `preparePresentationStateForScroll(to indexPath: IndexPath`? Usually don't encode type in Swift var names, IIRC.

File: Listable/Sources/ListView/ListView.swift [hunk 22]
```diff
@@ -154,6 +155,8 @@ public final class ListView : UIView
     {
         // Nothing right now.
     }
+
+    public var autoScrollAction : AutoScrollAction
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply Listable/Sources/ListView/ListView.swift:159]: Making this private - I did take a stab at removing the property, but it was pretty convoluted to pass it around right now. Would love to discuss with you separately from this PR though, since I could see us need to pass other "actions" like this that shouldn't actually be stored
